### ISSUE TITLE
Find closest language in whitelist

### ISSUE THREAD

**Issue by dcporter44** [i18next/i18next#1452]
Is there currently a way to load the closest language available?

For example: my app contains 2 languages, English (en) and Spanish Mexican (es-MX).

If a user loads my app with a browser language of es-ES, I would prefer for them to fallback to es-MX rather than en. This seems like common sense, but I can't find a way to do this.

`load: languageOnly` allows es-ES to fallback to es, but it does not search for closest match.

Is there a recommended way to do this?

**Comment by dcporter44** [comment]
By "find closest", I simply mean if the whole code (e.g. es-MX) is not found, then find any language that starts with es. If none found, then fallback to fallbackLng

**Comment by jamuhl** [comment]
By design...i18next has no knowledge of what languages you provide on the server -> it will try loading what was detected, what is the language only fallback, and the defined fallback...

you can use https://www.i18next.com/principles/fallback#language-fallback (see last sample with fallback as an object)

**Comment by dcporter44** [comment]
Isn't the `whitelist` there to tell i18next what languages are available on my server? What's the point of using a whitelist if I have to parse the language on my server anyway?

**Comment by jamuhl** [comment]
whitelist is for devs which do not like i18next making requests for languages not provided and getting a 404 network error

**Comment by jamuhl** [comment]
alternative have something like `en-US` -> `en` -> fallback

providing the language fallback should be sufficient in any case. So in your case:

es-ES -> es -> fallback
es-MX -> es -> fallback

**Comment by dcporter44** [comment]
Ok, I think I'm looking for functionality that does not exist.

Would you be open to me creating a PR for functionality called `useClosestInWhitelist: <boolean>`

For example, with `es-ES`:
1) Search whitelist for exactly `es-ES`
2) If `es-ES` does not exist in whitelist, search whitelist for `es`
3) If `es` does not exist in whitelist, search whitelist for any language that starts with `es-`
4) If cannot find anything, use fallback language.

Since my fallbackLng is `en`, in my case it would look like:
es-ES -> es -> es-MX -> en

**Comment by jamuhl** [comment]
1, 2, 4 already works

I really don't get why you need to look at es-MX if there is a full es

**Comment by dcporter44** [comment]
Because there might not be a full `es`. In my case, I only have 2 languages in my app: `en` and `es-MX`. Any browser language that starts with `es-`, I want to fallback to `es-MX` not `en`.

**Comment by jamuhl** [comment]
than rename es-MX to es ;)

**Comment by jamuhl** [comment]
if es-MX is the only variant of es you provide there is no sense in calling it es-MX

**Comment by dcporter44** [comment]
Yes, I could do that, but that seems like a poor workaround. What if in the future I want to add `es-CR`?

**Comment by dcporter44** [comment]
As a developer, I want to name my i18n files appropriately. I am going to end up writing custom logic here, and I'm sure that other developers will do the same thing in this situation.

**Comment by dcporter44** [comment]
I'm sure there are countless developers using this library that are writing the logic I wrote above to find the ideal language. I think it would help a lot of people if it was a simple config option.

**Comment by jamuhl** [comment]
in 10 years - you're the first...but yeah there might be some wanting this...still,l it's not needed complexity as there is already 100% more fallback options than in most other i18n frameworks out there -> where you can have only one file

Anyway...feel free to provide a PR - just if that gets too much complexity added I might not merge it

**Comment by jamuhl** [comment]
look up the code -> my guess needed to be added somewhere in the languageUtil where i18next resolves where to look it up...the load option currently is more a filter than an adding

**Comment by nitinkumar8553** [comment]
Hi @jamuhl,  before opening any issue I wanted to check something on my i18 config
so our application supports a list of languages(translation files) and if user language is not one of them then it will fallback to 'en'.
This setup was working but for an unsupported language, we start getting 404 error, to fix this we added whitelist in init object.
now with whitelist, we are having another problem. If the user language preference list is as given below.
1. Hi - not supported
2. de - supported
3. en - supported

then the language is being set to **de** instead of **en** which is the only fallback language.
the reason for this is coz de is already available in the whitelist so it is being added to the toResolveHierarchy and then the fallback language.  
code snippet  from toResolveHierarchy function
```

 fallbackCodes.forEach(function (fc) {
        if (codes.indexOf(fc) < 0) addCode(_this.formatLanguageCode(fc));
      });
      return codes;
```


here is my config.

const LANGUAGES = ['de', 'en', 'es', 'fr', 'it', 'ja', 'ko', 'pt-br', 'zh-cn', 'zh-tw'];
const i18nOptions = {
  backend: {
    loadPath: `${_contextRoot}/public/locales/{{lng}}/{{ns}}-resources.json`, // eslint-disable-line no-undef
  },
  saveMissing: false,
  nonExplicitWhitelist: false,
  debug: false,
  defaultNS: 'internetsvcs',
  fallbackLng: 'en',
  whitelist:LANGUAGES
  load: 'currentOnly',
  react: {
    nsMode: 'fallback',
    // we may want to use Suspense in the future
    useSuspense: false,
  },
  interpolation: {
    escapeValue: false, // not needed for react!!
  },
  lowerCaseLng: true,
  useCookie: false,
  detection: {
    order: ['querystring', 'navigator', 'htmlTag'],
  },
};

i18n
  .use(XHR)
  .use(LanguageDetector)
  .use(intervalPlural)
  .use(initReactI18next) // bind react-i18next to the instance
  .init(i18nOptions, () => {
    locI18next.init(i18n)('[data-i18n]'); // eslint-disable-line no-undef
    document.documentElement.lang = i18n.language || _language; // eslint-disable-line no-undef
  });

**Comment by jamuhl** [comment]
@nitinkumar8553 maybe I don't get it...the user has `de` in his browser language preferences and you support that language in your whitelist, but you expect the language to be set to `en` anyway?

**Comment by nitinkumar8553** [comment]
yes, I want to set language to en for any unsupported language irrespective of browser language preferences.

![Uploading Screenshot 2020-05-18 at 6.26.09 PM.png…]()


please find the attached screen shot.

**Comment by nitinkumar8553** [comment]
so the required is like this if the topmost language in the browser preference list is available in the whitelist then use it else fallback back to en only.

**Comment by nitinkumar8553** [comment]
@jamuhl  do let me know if you need more info on it.

**Comment by jamuhl** [comment]
@nitinkumar8553 this is a rather custom which...normally one would expect choosing the language preferred not - if it's not your top priority -> I send you to fallback even if I know you would prefer german?!? Isn't that a rather uncommon thing you like todo? idk...

If you really need that...you might have to build a own detector and inject that: https://github.com/i18next/i18next-browser-languageDetector#adding-own-detection-functionality based on https://github.com/i18next/i18next-browser-languageDetector/blob/master/src/browserLookups/navigator.js

**Comment by jamuhl** [comment]
Besides that...your "issue" is zero related to this one...not sure why people prefer "hijacking" other users issues to place there own problems?!?

**Comment by nitinkumar8553** [comment]
@jamuhl well I was not intending hijacking other's issue and I mentioned on top that before creating an issue I just wanted to check my config if this is correct as I didn't get any help from stack overflow questions.

FYI, it is indeed uncommon to not to switch to German, I added to test if fallback works correctly with my config.

I should have opened a new issue and discuss my problem over there.
somehow I'm able to achieve what I want by adding the below code.
{
lng: LANGUAGES.indexOf(userLanguage) < 0 ? 'en' : null,
}

**Comment by nitinkumar8553** [comment]
thanks for those links will give a try.

**Comment by dcporter44** [comment]
lol i've been hijacked

**Comment by jamuhl** [comment]
;) it's just something I see the last weeks...endless adds to old even closed issues - that are not really close to the issue created...it's a lot harder for us to keep track on those "side-issues"

**Comment by dcporter44** [comment]
@jamuhl i found a fun way to loop this back around. look at this guy's whitelist: ['de', 'en', 'es', 'fr', 'it', 'ja', 'ko', 'pt-br', 'zh-cn', 'zh-tw']

Notice 'pt-br' and no 'pt'. wouldn't it be nice if a 'pt' user would fallback to 'pt-br' instead of 'en'?

just saying 🙂

**Comment by nitinkumar8553** [comment]
@dcporter44 we are explicitly supporting these locale, 'pt-br', 'zh-cn', 'zh-tw'

**Comment by jamuhl** [comment]
why not fallback to latin?

**Comment by jamuhl** [comment]
In the end, it's the first detection...and you should always give the user an option to toggle it to what he likes

**Comment by dcporter44** [comment]
Are you implying that a man in portugal is just as likely to prefer Latin as he is to prefer brazillian portuguese

**Comment by nitinkumar8553** [comment]
as of now, we should fallback only to 'en', I did suggest but other folks on the team wanted to stick to this setting.

**Comment by jamuhl** [comment]
Hm...would guess a portuguese would prefer Spanish ;)...as a swissgerman would prefer french

**Comment by dcporter44** [comment]
@jamuhl  I get it. It's not "necessary", but I guarantee you a lot of people would use it if it was a config option.

**Comment by jamuhl** [comment]
You're PR will land soon...so that option will be available soon

**Comment by dcporter44** [comment]
🙂 coming soon

**Comment by nitinkumar8553** [comment]
> In the end, it's the first detection...and you should always give the user an option to toggle it to what he likes

yeah, maybe I would be testing in the wrong manner.

**Comment by KeelanM90** [comment]
Wow this is fresh, currently looking for this functionality too!

**Comment by dcporter44** [comment]
I decided this would make more sense as an option on i18next-browser-languagedetector:

PR: https://github.com/i18next/i18next-browser-languageDetector/pull/211

**Comment by jamuhl** [comment]
published in i18next-browser-languagedetector@4.2.0
